### PR TITLE
Upgrade to Serverless Framework v3 and the new design

### PR DESCRIPTION
**Pull request by mnapoli**
Quick preview:

https://user-images.githubusercontent.com/720328/140497195-253d27a3-a536-484f-a83f-dd4d1756f70d.mp4

Here is also a preview with `sls offline`, however with the v2 of `sls-offline` (so don't pay attention to the design of sls offline):

https://user-images.githubusercontent.com/720328/140497291-41c67076-6544-48f6-9dd2-74e10ae8cce3.mp4

Error:

image

I've opened this PR against master but I guess we'll need a `v3` branch with a beta release?

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -65,11 +84,17 @@ export class TypeScriptPlugin {
...
       },
-      'after:invoke:local:invoke': () => {
+      'after:invoke:local:invoke': async () => {
         if (this.options.watch) {
+          if (this.watchProgress) {
+            this.watchProgress.update('Watching for TypeScript changes')
+          }
           this.watchFunction()
-          this.serverless.cli.log('Waiting for changes...')
+          return new Promise(() => {
```
Comment (medikoo) [inline src/index.ts:95]: More accurate way would be to await on watchFunction, and fix `watchFunction` so it returns properly resolved promise.

Also it'll be great to have that patch in current major of a plugin (so best if it's proposed with other PR)

File: src/index.ts [hunk 2]
```diff
@@ -16,10 +16,29 @@ export class TypeScriptPlugin {
...
+  commands: Serverless.CommandsDefinition
+  watchProgress?: Serverless.Progress
 
-  constructor(serverless: Serverless.Instance, options: Serverless.Options) {
+  constructor(serverless: Serverless.Instance, options: Serverless.Options, utils) {
     this.serverless = serverless
     this.options = options
+    this.utils = utils
+
+    this.commands = {
```
Comment (medikoo) [inline src/index.ts:28]: This ideally should be propsed with another PR which can also land in current major of a plugin

**Review by medikoo — COMMENTED** [review]
Thanks @mnapoli. In general it looks good to me. I would just separate non-log related fixes to other PR's and publish them in scope of current major

File: src/index.ts (hunk 1 above)
Comment (mnapoli) [reply src/index.ts:95]: Yes, that's what I tried initially. But I think it's more complex than that, because `watchFunction()` calls `watchFiles()` that watches multiple files in parallel, and recursively re-spawns the CLI command in callbacks of each file changed. It seems like it isn't as simple as awaiting everything, given the command calls itself kind of recursively.

In other words, it seems to me like fixing this would require a more substantial refactoring/re-architecting of the way the plugin works, and that would be a bit too much for that pull request. If you manage to do that on master I can rebase on it of course, but I'd rather not attempt that here for lack of time, complexity, and also because this isn't really related to the redesign.

File: src/index.ts (hunk 2 above)
Comment (mnapoli) [reply src/index.ts:28]: 👍  I've opened https://github.com/serverless/serverless-plugin-typescript/pull/257

File: src/index.ts (hunk 1 above)
Comment (medikoo) [reply src/index.ts:95]: Ok, I'll give it a try against master and let you know

File: package.json [hunk 3]
```diff
@@ -62,12 +62,13 @@
...
   },
   "dependencies": {
+    "@serverless/utils": "pre-6",
     "fs-extra": "^7.0.1",
     "globby": "^10.0.2",
     "lodash": "^4.17.21"
   },
   "peerDependencies": {
-    "serverless": "2",
+    "serverless": "pre-3",
```
Comment (medikoo) [inline package.json:71]: I think we either should remove this setting (for eventual plugin pre-release case), or change it to `3` (assuming we want to merge and publish only after v3 is public).

`pre-3` cannot work as _peer_ setting, as it's  a _tag_ which always points a single publication (which changes with every push to v3 branch) and not a range of publications

Comment (mnapoli) [reply package.json:71]: The goal would be:

- v2 users should be able to install a compatible version of this plugin
- v3 beta users should be able to install a compatible version of this plugin (that uses the new design)
- when v3 is released, v3 users should be able to install a compatible version of this plugin

How can we make this work?

We'll have this issue on every plugin.

Comment (medikoo) [reply package.json:71]: As in case of this plugin, it's that we drop support for Framework v2, I think it needs to go with pre-release now, and `master` should stick to support for v2.

Once Framework v3 will be released, we can then officially issue a new major for this plugin.

Comment (mnapoli) [reply package.json:71]: ahhh ok I think I understand what you are trying to say: you are saying we release a beta version of this plugin, and v3-beta users have to install the beta version of this plugin explicitly?

Comment (medikoo) [reply package.json:71]: Yes, I think we cannot really release a public normal version, until there's a public normal version of Framework v3

**Comment by mnapoli** [comment]
FYI I removed the peer dependency.

**Review by medikoo — COMMENTED** [review]
Looks good, still it should not be merged to `master`, but rather `pre-3` pre-release branch

**Comment by mnapoli** [comment]
👍  I created the branch and changed the target of the PR.

I will let you take care of anything related to releases as you have a much clearer vision on this than me.

**Review by medikoo — APPROVED** [review]
Looks great 👍

**Comment by medikoo** [comment]
Published at `pre-3` tag

**Comment by mnapoli** [comment]
🎉
